File: homework/hw2/list.rb
#!/usr/bin/ruby 
#                 
# Author:  Abbas Moghtanei
# Date  :  09/18/09        
# Program Name: list.rb 
#                         
#********************************parse_range()****************************
def parse_range(str)

    if str == '-'
       from = 1; to = $NumLines
    elsif str == '$'
       from = to = $NumLines
    elsif !str.include?('-')          # like 87
       from = to = str
    else                              # it includes '-'
         p = str.index('-')           # find the position of '-'
         if p == 0                    # like -23
            from = 1
            to   = str[1...str.size]
         elsif p == str.size - 1      # like 23-
            from = str[0...p]
            to   = $NumLines
         else                         # like 23-156
            from = str[0,p]          
            to   = str[(p+1)...str.size]
         end
    end

    from = $NumLines if from == '$'
    to   = $NumLines if to   == '$'

    from = from.to_i if from.kind_of? String
    to   = to.to_i   if to.kind_of? String

    abort("Invalid range\n") if from <= 0 or to <= 0
    abort("Invalid range\n") if from > $NumLines or to > $NumLines

    return from,to
end
#******************************numlines()********************************
def numlines(filename)
  IO.readlines(filename).size
end
#******************************load_file()*******************************
def load_file(filename,from, to)
    temparr = []
    if from == 1 and to == $NumLines
       temparr = IO.readlines(filename)
       return temparr
    end
    
    n = 0 
    from, to = to, from if from > to
    open(filename, 'r') do |f|
       for i in 1..(from-1)
          f.gets
       end
       from.upto(to) { temparr[n] = f.gets; n += 1 }
    end
    temparr
end
#************************************************************************
txt = " <-n> line-range filename\n"
abort("usage: " + $0[2...$0.size] + "#{txt}")  if ARGV.size == 0
nflag = false
nflag = true if ARGV.delete("-n") == "-n"
abort("usage: " + $0[2...$0.size] + "#{txt}")  if ARGV.size == 0
if ARGV.size == 1
   line_range = nil
   filename = ARGV[0]
else
   line_range = ARGV[0]
   filename   = ARGV[1]
end

$NumLines = numlines(filename)
arr = []
line_range.nil? ? arr << "1-" : arr = line_range.split(',') 


arr.each do |r|
   from,to = parse_range(r)
   buf = load_file(filename, from, to)
   if from <= to
      for i in from..to
         if (nflag)
             printf("%4d %s", i, buf[i-1])
         else
             puts buf[i-1]
         end
      end
   else          
      while from >= to
         if (nflag)
             printf("%4d %s", from, buf[from-1])
         else
             puts buf[from-1]
         end
         from -= 1
      end
   end
end
puts

File: homework/hw2/list.pl
#!/usr/bin/perl -w
                 
# Author:   Steven lo
# Date  :  19 Sep 2009
# Program Name: list 
 
# Prints range(s) of text file lines to standard output.  If no range(s) specified, prints entire file.
 
# When invoked with -n option, prepends line(s) with line number(s).

#-----------------------------------------------------------------------------#

use strict;
 
use Getopt::Std;
 
die "usage: list <-n> <line-range> filename\n" if not @ARGV;

my @ARGV_reverse = reverse(@ARGV);

my $filename = $ARGV_reverse[0];

die "invalid filename:  $filename\n" if not (-e $filename);
die "$filename is not a text file\n" if not (-T $filename);

my $range_arg;

if ( (not $ARGV_reverse[1]) or ($ARGV_reverse[1] =~ m/^-[n]$/) )
{
  $range_arg = '1-$';
}
elsif ($ARGV_reverse[1] =~ m/^([-\$0-9]+,)*[-\$0-9]+$/)
{
  $range_arg = $ARGV_reverse[1];
}
else
{
  die "invalid range:  $ARGV_reverse[1]";
}

my %opts;

getopt('n', \%opts);

my %ranges;
my $range_num;

foreach my $range (split(m/,/, $range_arg, -1))
{
  $range_num++;

  $range =~ s/^-/1-/;
  $range =~ s/-$/-\$/;
  $range =~ s/^(\d+)$/$1-$1/;
  $range =~ s/^\$$/\$-\$/;

  my ($beg_line, $end_line) = split(m/-/, $range);

if ( ($beg_line eq '$' and $end_line ne '$') or ($end_line ne '$' and $beg_line gt $end_line) )
  {
    ($beg_line,$end_line) = ($end_line,$beg_line);

    $ranges{$range_num} = 
    {
      beg_line => $beg_line,
      end_line => $end_line,
      array => [],
      reverse => 1
    };
  }
  else
  {
    $ranges{$range_num} = 
    {
      beg_line => $beg_line,
      end_line => $end_line,
      array => [],
      reverse => 0
    };
  }
}
  
open(INPUT, '<', $filename) or die $!;

my $line;
my $prev_line;
my $linenum = 0;

if (exists $opts{n})
{
  while ($line = <INPUT>)
  {
    $linenum++;
  
    my $linenum_str = $linenum . '';

    $line = $linenum_str . " " x (8-length($linenum_str)) . $line;
  
    foreach my $key (keys %ranges)
    {
      if ( ($ranges{$key}{beg_line} ne '$' and ($ranges{$key}{beg_line} + 0) <= $linenum) and
           ($ranges{$key}{end_line} eq '$' or ($ranges{$key}{end_line} + 0) >= $linenum) )
      {
        push @{$ranges{$key}{array}}, $line;
      }
    }

    $prev_line = $line;
  }

  if ($prev_line)
  {
    foreach my $key (keys %ranges)
    {
      if ($ranges{$key}{beg_line} eq '$' and $ranges{$key}{end_line} eq '$')
      {
        push @{$ranges{$key}{array}}, $prev_line;
      }
    }
  }
}
else
{
  while (my $line = <INPUT>)
  {
    $linenum++;
  
    foreach my $key (keys %ranges)
    {
      if ( ($ranges{$key}{beg_line} ne '$' and ($ranges{$key}{beg_line} + 0) <= $linenum) and
           ($ranges{$key}{end_line} eq '$' or ($ranges{$key}{end_line} + 0) >= $linenum) )
      {
        push @{$ranges{$key}{array}}, $line;
      }
    }
  
    $prev_line = $line;
  }

  if ($prev_line)
  {
    foreach my $key (keys %ranges)
    {
      if ($ranges{$key}{beg_line} eq '$' and $ranges{$key}{end_line} eq '$')
      {
        push @{$ranges{$key}{array}}, $prev_line;
      }
    }
  }
}

foreach my $key (sort { $a <=> $b } (keys %ranges))
{
  if ($ranges{$key}{reverse} == 1)
  {
    print reverse(@{$ranges{$key}{array}});
  }
  else
  {
    print @{$ranges{$key}{array}};
  }
}
